src/lib/shim/src/tls.rs

```diff
@@ -441,33 +606,27 @@ impl<T, F> ShimTlsVar<T, F>
 where
     F: Fn() -> T,
 {
-    /// Create a variable that will be uniquely instantiated for each thread.
+    /// Create a variable that will be uniquely instantiated for each thread,
+    /// initialized with `f` on first access by each thread.
+    ///
     /// Typically this should go in a `static`.
     pub const fn new(f: F) -> Self {
         Self {
-            offset: LazyLock::const_new(OffsetInitializer::new()),
+            offset: LazyLock::const_new(OffsetInitializer::new::<T>()),
             f,
             _phantom: PhantomData,
         }
     }
 
-    const fn align() -> usize {
-        let align = core::mem::align_of::<ShimTlsVarStorage<T>>();
-        // Must be at least the alignment of `ShimThreadLocalStorage`.
-        // If this fails, we can increase `ShimThreadLocalStorage`'s alignment.
-        assert!(align <= 16);
-        align
-    }
-
     fn offset(&self) -> usize {
         *self.offset.force()
     }
 
     /// Access the inner value.
     ///
-    /// The returned wrapper can't be sent to or shared with other threads, to
-    /// allow the implementation to be extended in the future to deallocate
-    /// and/or reuse storage when a thread exits.
+    /// The returned wrapper can't be sent to or shared with other threads,
+    /// since the underlying storage is invalidated when the originating thread
+    /// calls [`unregister_current_thread`].
     pub fn get(&self) -> TLSVarRef<T> {
```

Comment:
I was thinking of moving the `exit` syscall *into* `unregister_current_thread`, and calling it something like `unregister_current_thread_and_exit` or just `exit_thread`.

The sequence would go:

* `sigprocmask` to block all* signals (see below), to ensure no signal handlers can run on the current thread
* unregister the thread
* `exit` the thread

I'll have to check what  happens with seccomp when `SIGSYS` is blocked; my hope is that it would kill the process, but I don't know for sure. We also can't block `SIGKILL` or `SIGSTOP` but neither of those are allowed to have signal handlers, either.

If we do that, I think that removes most of the concern here in `get`? I think the unregistration function could also no longer be `unsafe`

Reply:
Yeah I think this would be enough to make it safe. If we make a direct `exit` syscall and there aren't any signal handlers, I can't think of any other way that shim code could run between the unregistering/zeroing and the exit syscall.

Reply:
I guess it depends on if it would be considered UB for a dangling rust reference to exist even briefly (which I think is technically UB), but I don't think there's a solution to that other than not allowing `'static` references to thread-locals. But I think having the `unregister_current_thread` also exit would be better than having them separate.

Reply:
> I don't think there's a solution to that other than not allowing 'static references to thread-locals.

Actually I don't think this is right either, since you could still call `unregister_current_thread` while holding a non-static reference to a thread-local. Maybe a solution could be a global atomic counter, and when a `TLSVarRef` is created it increases and when it's dropped it decreases, and then the `unregister_current_thread` makes sure there are no references. But I don't know what the performance implication is, and I don't think that's needed for this PR.

Reply:
I *think* the counter could just be a `Cell` in `ShimThreadLocalStorage`, but yeah I'll probably leave that for a separate PR.